File: leetcode/problems_141_linked-list-cycle.c
#include <stdbool.h> 
#include <stdio.h>

struct ListNode {
  int val;
  struct ListNode *next;
};

bool hasCycle(struct ListNode *head) {
  if (head == NULL) {
    return false;
  }

  struct ListNode *slow = head;
  struct ListNode *fast = head;
  
  do {
    if (fast->next == NULL || fast->next->next == NULL) {
      return false;
    } else {
      fast = fast->next->next;
    }

    slow = slow->next;
  } while(slow != fast);

  return true;
}

int main() {
  return 0;
}

File: leetcodecn/problems_141_linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        while (head != null) {
            if (head.val == 100001) {
                return true;
            } else {
                head.val = 100001;
                head = head.next;
            }
        }
        return false;
    }


    public boolean hasCycle1(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode fast = head;
        ListNode slow = head;
        while (fast != null && fast.next != null) {
            fast =  fast.next.next;
            slow = slow.next;
            if (fast == slow) return true;
        }

        return false;
    }
}
